### DISCUSSION TITLE
`serde_as` cannot be cfg-gated

### DISCUSSION THREAD

**Discussion by chipsenkbeil** [jonasbb/serde_with#331] [Q&A] [answered]
I've seen some examples of using `serde_as` for explicitly-defined size ranges, but I'm struggling to figure out how to apply to const generics, if it's even possible.

```rust
use serde::{Serialize, Deserialize};
use serde_with::serde_as;

#[serde_as]
#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
struct FixedMatrix<T, const ROW: usize, const COL: usize> {
    #[serde_as(as = "[[_; ROW]; COL]")]
    data: [[T; ROW]; COL],
}
```

Was browsing through https://github.com/serde-rs/serde/issues/1937 and saw mention of:

```rust
#[serde_with::serde_as]
#[derive(Serialize)]
struct Foo {
    #[serde_as(as = "[_; 2]")]
    bar: [String; 2],
    #[serde_as(as = "Vec<[_; 5]>")]
    foobar: Vec<[u8; 5]>,
}
```

**Comment by jonasbb** [comment, marked as answer]
Since you use `serde_as` on a field with a generic, you need to manually specify the correct trait bounds for de-/serializing type `T`.

```rust
use serde::{Serialize, Deserialize};
use serde_with::serde_as;

#[serde_as]
#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(bound(serialize = "T: Serialize", deserialize = "T: Deserialize<'de>"))]
struct FixedMatrix<T, const ROW: usize, const COL: usize> {
    #[serde_as(as = "[[_; ROW]; COL]")]
    data: [[T; ROW]; COL],
}
```

Applying any of the serde attributes `with`, `serialize_with`, or `deserialize_with` on generics changes how serde infers the trait bounds of the resulting implementations. By default, serde adds `T: Serialize` as a requirement to the `Serialize` implementation, but not if those attributes are specified.

The same applies to `serde_as`. You need to tell serde that `T` can be de-/serialized by specifying the correct trait bounds. Then it will compile without problems.

---

Your attribute would work as is, if a normal type is used in the array, e.g., `[[u8; ROW]; COL]`.

```rust
#[serde_as(as = "[[_; ROW]; COL]")]
data: [[u8; ROW]; COL],
```

**Reply by chipsenkbeil** [reply]
@jonasbb it makes sense and it's what I had to do (setting bounds) after rolling my own serialize/deserialize functions. For some reason, I still can't seem to get it to work with `serde_as` as it complains about both traits `Serialize` and `Deserialize<'de>` not being implemented for `[[T; COL]; ROW]` as a whole. I've read through how `serde_as` does it and rolled my own serialize functions like [`serialize_table_array`](https://github.com/chipsenkbeil/memtable-rs/blob/6cb499fd0927d4dde9be34c2bde0cbc1b121c09c/memtable-core/src/impls/utils/ser/table_array.rs#L6). Would obviously love to not have to maintain code like this myself.

Tried adding bounds both at the container and field level, but doesn't seem to affect the error I'm seeing. Don't want to take up your time here as I was working on this quite late and may have easily skipped over something obvious like with the trait bounds earlier. Mostly documenting all I know for other folks that swing by in case they hit a similar challenge.

# serde_as 

```toml
serde = { version = "1.0.117", features = ["derive"] }
serde_with = "1.9.4"
```

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/2481802/123551902-6efb4200-d739-11eb-885c-7a5d4402f47d.png">

# My Implementation

```rust
#[derive(Clone, Debug, Eq, PartialEq)]
#[cfg_attr(feature = "serde-1", derive(serde::Serialize, serde::Deserialize))]
#[repr(transparent)]
pub struct FixedTable<T: Default, const ROW: usize, const COL: usize>(
    #[cfg_attr(
        feature = "serde-1",
        serde(
            bound(
                serialize = "T: serde::Serialize",
                deserialize = "T: serde::Deserialize<'de>"
            ),
            serialize_with = "utils::serialize_table_array",
            deserialize_with = "utils::deserialize_table_array"
        )
    )]
    [[T; COL]; ROW],
);
```

### serialize_table_array

```rust
use serde::ser;

/// Workaround for https://github.com/serde-rs/serde/issues/1937
pub fn serialize_table_array<S, T, const ROW: usize, const COL: usize>(
    value: &[[T; COL]; ROW],
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: ser::Serializer,
    T: ser::Serialize,
{
    use ser::SerializeTuple;
    let mut tup = serializer.serialize_tuple(ROW * COL)?;
    for row in 0..ROW {
        for col in 0..COL {
            tup.serialize_element(&value[row][col])?;
        }
    }
    tup.end()
}
```

### deserialize_table_array

```rust
use super::default_table_array;
use serde::de;

/// Workaround for https://github.com/serde-rs/serde/issues/1937
pub fn deserialize_table_array<'de, D, T, const ROW: usize, const COL: usize>(
    deserializer: D,
) -> Result<[[T; COL]; ROW], D::Error>
where
    D: de::Deserializer<'de>,
    T: de::Deserialize<'de> + Default,
{
    deserializer.deserialize_tuple(ROW * COL, TableArrayVisitor::<T, ROW, COL>::default())
}

#[derive(Default)]
struct TableArrayVisitor<T, const ROW: usize, const COL: usize> {
    _marker: core::marker::PhantomData<T>,
}

impl<'de, T, const ROW: usize, const COL: usize> de::Visitor<'de> for TableArrayVisitor<T, ROW, COL>
where
    T: de::Deserialize<'de> + Default,
{
    type Value = [[T; COL]; ROW];

    /// Format a message stating we expect an array of size `N`
    fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
        write!(formatter, "an array of size {}", ROW * COL)
    }

    /// Process a sequence into a table array
    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: de::SeqAccess<'de>,
    {
        let mut table = default_table_array::<T, ROW, COL>();
        let mut cnt = 0;

        for row in 0..ROW {
            for col in 0..COL {
                let next = seq.next_element::<T>()?;

                // Assign the element to our array if we have it
                if let Some(x) = next {
                    table[row][col] = x;
                    cnt += 1;

                // Otherwise, we have a bad sequence
                } else {
                    return Err(de::Error::invalid_length(cnt, &self));
                }
            }
        }

        // If we still have more elements, there's a problem
        if seq.next_element::<T>()?.is_some() {
            Err(de::Error::invalid_length(cnt + 1, &self))

        // Otherwise, we're good to go
        } else {
            Ok(table)
        }
    }
}
```

**Reply by jonasbb** [reply]
@chipsenkbeil Good to hear that you found a solution. One note about your implementation: the `#[serde(bound( ... ))]` part needs to be on the struct/enum, not on the fields.

I think I found the problem. You are cfg-gating the `serde_as` field attribute. This `cfg_attr` is still present when the `serde_as` proc-macro gets run. The proc-macro does not recognize the `cfg_attr` and simply ignores it. When the serde derives run, they do not see any attributes on the field and fail to serialize it, due to missing const generics support.

I don't think it is currently possible to eagerly expand `cfg_attr` in a proc-macro, since IIRC the expansion order is proc-macros, `cfg_attr`, derive-macros.
It could be solved by distributing a set of `serde_derive` like macros which directly understand the `serde_as` syntax.

---

I noticed your `serialize_table_array` does something different from `[[T; COL]; ROW]`, namely it serializes into a single sequence, but the arrays are multiple nested sequences. Using `serde_with` will not help here.

**Reply by chipsenkbeil** [reply]
@jonasbb I just flattened it because that seemed simpler to do, not that I require it to be done that way in practice. :)

What's interesting is that I'm able to use the cfg-gating for another [struct with a HashMap<_, T>](https://github.com/chipsenkbeil/memtable-rs/blob/6cb499fd0927d4dde9be34c2bde0cbc1b121c09c/memtable-core/src/impls/mem.rs#L13). But you're right in that after you mentioned that I'm surprised it worked.

**Reply by jonasbb** [reply]
Have you tried running that piece of code? It will compile if `Position` position is serializable. But it will try to serialize it as a HashMap not as a list of tuples.

**Reply by chipsenkbeil** [reply]
@jonasbb I've not written any test to verify that serde serialization/deserialization of the table works. It's something I should do as I was bitten by the HashMap with custom key when serializing to `serde_json` in the past. It's how I found your fantastic library 😄 

Cleaning up some parts of the library today, so I'll get around to testing it and seeing what happens. All I know so far is that it compiles w/ and w/o `serde` being incorporated, to my knowledge.

**Reply by jonasbb** [reply]
There is some work in rustc to have a `#[cfg_eval]` attribute, which will eagerly evaluate all cfgs before the proc macros. There is a tracking issue https://github.com/rust-lang/rust/issues/82679 and a stabilization PR https://github.com/rust-lang/rust/pull/83824.

It might be possible for `serde_as` to see through the `cfg_attr` and still do all the `serde_as` transformations. Basically, transform `#[cfg_attr(foobar, serde_as(as = "..."))]` into `#[cfg_attr(foobar, serde(with = "..."))]`.
